x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.text.ParseException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.elasticsearch.test.TestMatchers.hasStatusCode;
+import static org.hamcrest.Matchers.arrayWithSize;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+import static org.hamcrest.Matchers.hasKey;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+
+public class JwtRestIT extends ESRestTestCase {
+
+    private static final Optional<String> VALID_SHARED_SECRET = Optional.of("test-secret");
+    private static Path httpCertificateAuthority;
+    private TestSecurityClient adminSecurityClient;
+
+    @BeforeClass
+    public static void findTrustStore() throws Exception {
+        final URL resource = JwtRestIT.class.getResource("/ssl/ca.crt");
+        if (resource == null) {
+            throw new FileNotFoundException("Cannot find classpath resource /ssl/ca.crt");
+        }
+        httpCertificateAuthority = PathUtils.get(resource.toURI());
+    }
+
+    @Override
+    protected String getProtocol() {
+        // Because this QA project uses https
+        return "https";
+    }
+
+    @Override
+    protected Settings restAdminSettings() {
+        String token = basicAuthHeaderValue("admin_user", new SecureString("admin-password".toCharArray()));
+        return Settings.builder().put(ThreadContext.PREFIX + ".Authorization", token).put(restSslSettings()).build();
+    }
+
+    @Override
+    protected Settings restClientSettings() {
+        return Settings.builder().put(super.restClientSettings()).put(restSslSettings()).build();
+    }
+
+    private Settings restSslSettings() {
+        return Settings.builder().put(CERTIFICATE_AUTHORITIES, httpCertificateAuthority).build();
+    }
+
+    protected TestSecurityClient getAdminSecurityClient() {
+        if (adminSecurityClient == null) {
+            adminSecurityClient = new TestSecurityClient(adminClient());
+        }
+        return adminSecurityClient;
+    }
+
+    /**
+     * Tests against realm "jwt1" in build.gradle
+     * This realm
+     * - use the "sub" claim as the principal
+     * - performs role mapping
+     * - supports RSA signed keys
+     * - has no client authentication
+     */
+    public void testAuthenticateWithRsaSignedJWTAndRoleMapping() throws Exception {
+        final String principal = randomPrincipal();
+        final List<String> roles = randomRoles();
+        final String roleMapping = createRoleMapping(roles, """
+            { "all": [
+                { "field": { "realm.name": "jwt1" } },
+                { "field": { "username": "%s" } }
+            ] }
+            """.formatted(principal));
```

Comment:
I added tests for role mapping on groups and `jwt_claim_sub`.
I'm not sure we need test the _role mapping_ on groups, but we needed something that tested the extraction of groups from the claims, so that seemed like the simplest approach.